### ISSUE TITLE
Add explicit "enum type" to VRL

### ISSUE THREAD

**Issue by JeanMertz** [timberio/vector#8482]
We've been introducing functions that take an enum variant as an argument.

For example:

```coffee
parse_nginx_log(.message, format: "combined")
```

The `format` argument has to be either `"combined"` or `"error"`, and has to be a string literal. This has been awkwardly implemented and can be confusing to users. It also results in less-than-ideal error messages if people try to pass in non-literal string values (e.g. through an event field).

I've been pondering a solution, and although having full-blown enum support (both defining and referencing) can be powerful, I also believe it doesn't fit the simplicity we want to achieve with VRL.

Instead, I think we should add a new enum type that can only be referenced as a literal, similar to how we have a timestamp type.

Specifically, I propose we add `e'...'` as a type for an enum variant literal.

Given this function signature:

```coffee
foo(timestamp: Timestamp, variant: Enum)
```

You'd call this function like so:

```coffee
foo(timestamp: t'2021-01-01T00:00:00Z', variant: e'my_variant')
```

This would only compile if `variant` is passed an enum type (`e'...'`) _and_ if the provided enum variant is accepted by the function signature.

Function signatures would define the set of enum variants they accept, which the compiler can then automatically check (and provide useful error messages indicating which enum variants are expected and which one was provided). This is an improvement over the current situation because functions now have to build their own (non-ideal) error messages when dealing with enum-like strings, given that we don't actually have support for enums.

All of this would also bubble up into the Cue docs, which would result in better auto-generated function signatures and a list of enum variants supported for a specific parameter.

**Comment by hhromic** [comment]
This is a great idea! I'm curious if the enum type would be scoped by their application.

For example the `parse_nginx_log()` function would only accept a particular Enum which may or may not be the same as for other functions. In this case, how the `e'...'` type would work? By the context where it is used?

For example:
```coffee
parse_nginx_log(message, format: e'combined')
parse_key_value(message, whitespace: e'lenient')
```
These two enum values belong to two totally different enums.

**Comment by JeanMertz** [comment]
> These two enum values belong to two totally different enums.

That's correct. However, the enum value wouldn't be validated until the function call is compiled. So, the variant you provide, is context-aware, and can be valid for one function call but invalid for another.

I'll say, I'm not _entirely_ sold on the idea yet (or the implementation details). It does solve a specific issue we see in the wild, but there might be other solutions to that problem, we'll have to see.
